File: internal/nodeapproval/fingerprint.go
// Package nodeapproval implements the opt-in signed-preflight PoC.
package nodeapproval

import (
	"bytes"
	"encoding/json"
	"fmt"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/crypto"
	"math/big"
	"privacy-proxy/internal/tracer"
	"sort"
	"strconv"
	"strings"
)

type object = map[string]any

func obj(v any) object {
	m, _ := v.(map[string]any)
	if m == nil {
		return object{}
	}
	return m
}
func str(v any) string { s, _ := v.(string); return s }
func word(v any) (string, error) {
	s := strings.TrimPrefix(str(v), "0x")
	if len(s) > 64 {
		return "", fmt.Errorf("oversize word")
	}
	for _, c := range s {
		if !strings.ContainsRune("0123456789abcdefABCDEF", c) {
			return "", fmt.Errorf("invalid word")
		}
	}
	return "0x" + strings.Repeat("0", 64-len(s)) + strings.ToLower(s), nil
}
func field(v object, k, def string) string {
	if s, ok := v[k].(string); ok {
		return strings.ToLower(s)
	}
	return def
}

func normalizeCall(v object, parent string, count *int, trace *tracer.TraceResult, depth int) (object, error) {
	*count++
	if *count > 128 {
		return nil, fmt.Errorf("more than 128 calls")
	}
	kind := str(v["type"])
	if kind != "CALL" && kind != "STATICCALL" && kind != "DELEGATECALL" && kind != "CALLCODE" && kind != "CREATE" && kind != "CREATE2" && kind != "SELFDESTRUCT" {
		return nil, fmt.Errorf("unsupported call %q", kind)
	}
	value, err := word(v["value"])
	if err != nil {
		return nil, err
	}
	trace.HasCreate = trace.HasCreate || kind == "CREATE"
	trace.HasCreate2 = trace.HasCreate2 || kind == "CREATE2"
	to, from := field(v, "to", ""), field(v, "from", "")
	// Pinned revm-inspectors derives SELFDESTRUCT metadata from journal entries.
	// Cancun self-to-self on an existing account has no entry: its marker has
	// zero caller, no destination/value and succeeds. It denotes a no-op on parent.
	zero, _ := word(nil)
	if kind == "SELFDESTRUCT" && to == "" && from == (common.Address{}).Hex() && value == zero && v["error"] == nil && parent != "" {
		from, to = parent, parent
	}

	failedCreate := (kind == "CREATE" || kind == "CREATE2") && v["error"] != nil && to == ""
	if (!common.IsHexAddress(to) && !failedCreate) || !common.IsHexAddress(from) {
		return nil, fmt.Errorf("invalid call address")
	}
	storage := to
	if kind == "SELFDESTRUCT" {
		storage = from
	}
	if kind == "DELEGATECALL" || kind == "CALLCODE" {
		storage = parent
		if storage == "" {
			return nil, fmt.Errorf("root delegate")
		}
	}
	trace.CallTargets = append(trace.CallTargets, tracer.CallTarget{Type: kind, From: from, To: to, Depth: depth})
	children := []any{}
	if list, ok := v["calls"].([]any); ok {
		for _, c := range list {
			n, e := normalizeCall(obj(c), storage, count, trace, depth+1)
			if e != nil {
				return nil, e
			}
			children = append(children, n)
		}
	}
	logs := v["logs"]
	if logs == nil {
		logs = []any{}
	}
	return object{"type": kind, "from": from, "to": to, "storageAddress": storage, "input": field(v, "input", "0x"), "output": field(v, "output", "0x"), "failed": v["error"] != nil, "value": value, "calls": children, "logs": logs}, nil
}

// Fingerprint uses the same canonical JSON/domain as the producer module.
func Fingerprint(calls, pre, diff object) (common.Hash, *tracer.TraceResult, error) {
	trace := &tracer.TraceResult{}
	count := 0
	c, err := normalizeCall(calls, "", &count, trace, 0)
	if err != nil {
		return common.Hash{}, nil, err
	}
	accounts := []any{}
	created := map[string]bool{}
	for _, t := range trace.CallTargets {
		if t.Type == "CREATE" || t.Type == "CREATE2" {
			created[t.To] = true
		}
	}
	addresses := map[string]bool{}
	for a := range pre {
		addresses[a] = true
	}
	for a := range obj(diff["pre"]) {
		addresses[a] = true
	}
	for a := range obj(diff["post"]) {
		addresses[a] = true
	}
	ordered := make([]string, 0, len(addresses))
	for a := range addresses {
		ordered = append(ordered, a)
	}
	sort.Strings(ordered)
	for _, a := range ordered {
		if !common.IsHexAddress(a) {
			return common.Hash{}, nil, fmt.Errorf("invalid account address")
		}
		before := obj(pre[a])
		changedPre, hasPre := obj(diff["pre"])[a]
		changedPost, hasPost := obj(diff["post"])[a]
		deleted := hasPre && !hasPost
		post := obj(changedPost)
		codeBefore := field(before, "code", "0x")
		codeAfter := field(post, "code", codeBefore)
		if deleted {
			codeAfter = "0x"
		}
		balanceBefore, e := number(before["balance"])
		if e != nil {
			return common.Hash{}, nil, e
		}
		balanceAfter := new(big.Int).Set(balanceBefore)
		if v, ok := post["balance"]; ok {
			balanceAfter, e = number(v)
		}
		if deleted {
			balanceAfter = new(big.Int)
		}
		if e != nil {
			return common.Hash{}, nil, e
		}
		delta := new(big.Int).Sub(balanceAfter, balanceBefore).String()
		slotsBefore := obj(before["storage"])
		slotsPre, slotsPost := obj(obj(changedPre)["storage"]), obj(post["storage"])
		slots := map[string]bool{}
		for k := range slotsBefore {
			slots[k] = true
		}
		for k := range slotsPre {
			slots[k] = true
		}
		for k := range slotsPost {
			slots[k] = true
		}
		contract := codeBefore != "0x" || codeAfter != "0x" || len(slots) != 0 || created[strings.ToLower(a)]
		if !contract && delta == "0" {
			continue
		}
		nb, na := "0", "0"
		if contract {
			n, e := number(before["nonce"])
			if e != nil {
				return common.Hash{}, nil, e
			}
			nb = n.String()
			na = nb
			if v, ok := post["nonce"]; ok {
				n, e = number(v)
				if e != nil {
					return common.Hash{}, nil, e
				}
				na = n.String()
			}
			if deleted {
				na = "0"
			}
		}
		storage := []any{}
		for k := range slots {
			slot, e := word(k)
			if e != nil {
				return common.Hash{}, nil, e
			}
			bv, e := word(slotsBefore[k])
			if e != nil {
				return common.Hash{}, nil, e
			}
			av := bv
			_, p := slotsPre[k]
			_, q := slotsPost[k]
			if deleted || p || q {
				av, e = word(slotsPost[k])
				if e != nil {
					return common.Hash{}, nil, e
				}
			}
			storage = append(storage, object{"slot": slot, "before": bv, "after": av})
		}
		sort.Slice(storage, func(i, j int) bool { return str(obj(storage[i])["slot"]) < str(obj(storage[j])["slot"]) })
		accounts = append(accounts, object{"address": strings.ToLower(a), "codeBefore": crypto.Keccak256Hash(common.FromHex(codeBefore)).Hex(), "codeAfter": crypto.Keccak256Hash(common.FromHex(codeAfter)).Hex(), "nonceBefore": nb, "nonceAfter": na, "balanceDelta": delta, "storage": storage})
	}
	var buffer bytes.Buffer
	enc := json.NewEncoder(&buffer)
	enc.SetEscapeHTML(false)
	if err = enc.Encode(object{"calls": c, "accounts": accounts}); err != nil {
		return common.Hash{}, nil, err
	}
	data := bytes.TrimSuffix(buffer.Bytes(), []byte{'\n'})
	if len(data) > 1048576 {
		return common.Hash{}, nil, fmt.Errorf("fingerprint exceeds 1 MiB")
	}
	return crypto.Keccak256Hash([]byte("OPS_EXECUTION_V2\x00"), data), trace, nil
}

// RPC nonce fields can be JSON numbers; decoding with UseNumber preserves uint64.
func number(v any) (*big.Int, error) {
	if v == nil {
		return new(big.Int), nil
	}
	var s string
	switch n := v.(type) {
	case string:
		s = n
	case json.Number:
		s = string(n)
	case float64:
		if n < 0 || n > 1<<53 || n != float64(uint64(n)) {
			return nil, fmt.Errorf("inexact integer")
		}
		s = strconv.FormatUint(uint64(n), 10)
	default:
		return nil, fmt.Errorf("invalid integer %T", v)
	}
	base := 10
	if strings.HasPrefix(s, "0x") {
		s = s[2:]
		base = 16
	}
	n, ok := new(big.Int).SetString(s, base)
	if !ok || n.Sign() < 0 || n.BitLen() > 256 {
		return nil, fmt.Errorf("invalid uint256")
	}
	return n, nil
}

File: poc/reth-signed-approvals/src/fingerprint.rs
//! V2 reference encoder: call/value tree, code/storage/nonces and application balance deltas.
//! RPC preflight has no protocol fee charges on the pinned Reth version.
use alloy_primitives::{B256, U256, hex, keccak256};
use serde_json::{Value, json};
use std::collections::BTreeSet;

pub const DOMAIN: &[u8] = b"OPS_EXECUTION_V2\0";

fn word(v: Option<&Value>) -> Result<String, String> {
    let s = v
        .and_then(Value::as_str)
        .unwrap_or("0x0")
        .trim_start_matches("0x");
    if s.len() > 64 || !s.bytes().all(|b| b.is_ascii_hexdigit()) {
        return Err("bad storage word".into());
    }
    Ok(format!("0x{:0>64}", s.to_lowercase()))
}

fn call(v: &Value, parent_storage: Option<&str>, count: &mut usize) -> Result<Value, String> {
    *count += 1;
    if *count > 128 {
        return Err("more than 128 calls".into());
    }
    let kind = v["type"].as_str().ok_or("missing call type")?;
    if ![
        "CALL",
        "STATICCALL",
        "DELEGATECALL",
        "CALLCODE",
        "CREATE",
        "CREATE2",
        "SELFDESTRUCT",
    ]
    .contains(&kind)
    {
        return Err(format!("unsupported call: {kind}"));
    }
    let failed_create =
        ["CREATE", "CREATE2"].contains(&kind) && v.get("error").is_some_and(|e| !e.is_null());
    let mut from = v["from"].as_str().ok_or("missing caller")?.to_lowercase();
    let noop_self = kind == "SELFDESTRUCT"
        && v["to"].is_null()
        && from == "0x0000000000000000000000000000000000000000"
        && word(v.get("value"))? == word(None)?
        && v["error"].is_null()
        && parent_storage.is_some();
    let to = if noop_self {
        from = parent_storage.unwrap().to_lowercase();
        from.clone()
    } else {
        v["to"]
            .as_str()
            .or(failed_create.then_some(""))
            .ok_or("missing call destination")?
            .to_lowercase()
    };
    let storage = if kind == "DELEGATECALL" || kind == "CALLCODE" {
        parent_storage.ok_or("root delegate")?
    } else if kind == "SELFDESTRUCT" {
        &from
    } else {
        &to
    };
    let children = v
        .get("calls")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .map(|c| call(c, Some(storage), count))
        .collect::<Result<Vec<_>, _>>()?;
    Ok(json!({"type":kind,"from":from,
        "to":to,"storageAddress":storage,"input":v["input"].as_str().unwrap_or("0x").to_lowercase(),
        "output":v["output"].as_str().unwrap_or("0x").to_lowercase(),
        "failed":v.get("error").is_some_and(|e| !e.is_null()),"value":word(v.get("value"))?,"calls":children,
        "logs":v.get("logs").cloned().unwrap_or(json!([]))}))
}

pub fn number(v: &Value) -> Result<U256, String> {
    match v {
        Value::Null => Ok(U256::ZERO),
        Value::String(s) => s.parse::<U256>().map_err(|e| e.to_string()),
        Value::Number(n) => n.as_u64().map(U256::from).ok_or("invalid uint64".into()),
        _ => Err("invalid integer".into()),
    }
}

pub fn balance_delta(before: U256, after: U256) -> String {
    if after >= before {
        (after - before).to_string()
    } else {
        format!("-{}", before - after)
    }
}

pub fn canonical(calls: &Value, pre: &Value, diff: &Value) -> Result<Value, String> {
    let calls = call(calls, None, &mut 0)?;
    let mut accounts = Vec::new();
    let mut created = BTreeSet::new();
    fn creations(v: &Value, addresses: &mut BTreeSet<String>) {
        if matches!(v["type"].as_str(), Some("CREATE" | "CREATE2"))
            && let Some(to) = v["to"].as_str()
        {
            addresses.insert(to.to_lowercase());
        }
        for child in v["calls"].as_array().into_iter().flatten() {
            creations(child, addresses);
        }
    }
    creations(&calls, &mut created);
    let mut addresses = BTreeSet::new();
    for map in [pre, &diff["pre"], &diff["post"]] {
        if let Some(map) = map.as_object() {
            addresses.extend(map.keys());
        }
    }
    for address in addresses {
        let account = &pre[address];
        let before_diff = &diff["pre"][address];
        let post = &diff["post"][address];
        let deleted = !before_diff.is_null() && post.is_null();
        let code_before = account["code"].as_str().unwrap_or("0x");
        let code_after = if deleted {
            "0x"
        } else {
            post["code"].as_str().unwrap_or(code_before)
        };
        let balance_before = number(&account["balance"])?;
        let balance_after = if deleted {
            U256::ZERO
        } else if post["balance"].is_null() {
            balance_before
        } else {
            number(&post["balance"])?
        };
        let delta = balance_delta(balance_before, balance_after);
        let mut slots = BTreeSet::new();
        for map in [
            &account["storage"],
            &before_diff["storage"],
            &post["storage"],
        ] {
            if let Some(map) = map.as_object() {
                slots.extend(map.keys());
            }
        }
        let contract = code_before != "0x"
            || code_after != "0x"
            || !slots.is_empty()
            || created.contains(&address.to_lowercase());
        if !contract && delta == "0" {
            continue;
        }
        let nonce_before = if contract {
            number(&account["nonce"])?
        } else {
            U256::ZERO
        };
        let nonce_after = if !contract || deleted {
            U256::ZERO
        } else if post["nonce"].is_null() {
            nonce_before
        } else {
            number(&post["nonce"])?
        };
        let mut storage = Vec::new();
        for slot in slots {
            let before = account["storage"].get(slot);
            let changed = deleted
                || before_diff["storage"].get(slot).is_some()
                || post["storage"].get(slot).is_some();
            let after = if changed {
                post["storage"].get(slot)
            } else {
                before
            };
            storage.push(json!({"slot":word(Some(&json!(slot)))?,"before":word(before)?,"after":word(after)?}));
        }
        storage.sort_by(|a, b| a["slot"].as_str().cmp(&b["slot"].as_str()));
        accounts.push(json!({"address":address.to_lowercase(),"codeBefore":keccak256(hex::decode(code_before).map_err(|e|e.to_string())?),"codeAfter":keccak256(hex::decode(code_after).map_err(|e|e.to_string())?),"nonceBefore":nonce_before.to_string(),"nonceAfter":nonce_after.to_string(),"balanceDelta":delta,"storage":storage}));
    }
    Ok(json!({"calls":calls,"accounts":accounts}))
}

pub fn fingerprint(calls: &Value, pre: &Value, diff: &Value) -> Result<B256, String> {
    profiled_fingerprint(calls, pre, diff, &mut crate::profile::Timer::new(false))
}

pub fn profiled_fingerprint(
    calls: &Value,
    pre: &Value,
    diff: &Value,
    timer: &mut crate::profile::Timer,
) -> Result<B256, String> {
    let canonical = canonical(calls, pre, diff)?;
    timer.lap(6);
    let bytes = serde_json::to_vec(&canonical).map_err(|e| e.to_string())?;
    drop(canonical);
    timer.lap(7);
    if bytes.len() > 1_048_576 {
        return Err("fingerprint exceeds 1 MiB".into());
    }
    let mut input = DOMAIN.to_vec();
    input.extend(bytes);
    let hash = keccak256(input);
    timer.lap(8);
    Ok(hash)
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn shared_go_rust_golden() {
        let v: Value = serde_json::from_str(include_str!(
            "../../../internal/nodeapproval/testdata/fingerprint.json"
        ))
        .unwrap();
        assert_eq!(
            fingerprint(&v["calls"], &v["pre"], &v["diff"])
                .unwrap()
                .to_string(),
            v["expected"].as_str().unwrap()
        );
    }
}
